### PR TITLE
Low precision groupnorm

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Adds low precision groupnorm. This is modeled after low precision layernorm and is useful for stable diffusion. We've seen about +5% throughput and memory savings with this with no significant impact to loss curves.

# What issue(s) does this change relate to?

[CO-1794](https://mosaicml.atlassian.net/browse/CO-1794)

[CO-1794]: https://mosaicml.atlassian.net/browse/CO-1794?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by dskhudia** [comment]
Thought on combining this with lowprecisonlayernorm and calling that lowPrecisionNorm?

**Comment by mvpatel2000** [comment]
> Thoughs on combining this with lowprecisonlayernorm and calling that lowPrecisionNorms?

I would prefer to keep separate because we don't have guarantees of convergence with this, and there might be a case where one applies but another does not. While we have never observed issues with this, it is possible there are problems given AMP does not do this by default

**Comment by dskhudia** [comment]
We can allow turning on/off each type based on an option.

**Comment by nik-mosaic** [comment]
Is this scriptable? Can you export a model with LowPrecisionGroupnorm using torch.jit.script?

File: tests/algorithms/test_low_precision_groupnorm.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def assert_is_lpgn_instance(model):
+    # ensure that within the entire model, no PyTorch LayerNorm exists, and at least one APEX FLN does.
+    assert model.modules is not None, 'model has .modules method'
+    for module_class in model.modules():
+        if isinstance(module_class, LayerNorm):
+            assert isinstance(
+                module_class,
+                LPGroupNorm), 'A standard torch.nn.LayerNorm should not be found in the model after surgery is applied.'
```
Comment (dblalock) [inline tests/algorithms/test_low_precision_groupnorm.py:23]: less awkward line splitting?
```suggestion
            message = ('A standard torch.nn.LayerNorm should not ' +
                       'be found in the model after surgery is applied.')
            assert isinstance(module_class, LPGroupNorm), message
```

File: tests/algorithms/test_low_precision_groupnorm.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    assert any(
+        isinstance(module_class, LPGroupNorm) for module_class in model.modules()
+    ), 'composer.algorithms.low_precision_groupnorm.low_precision_groupnorm.LPGroupNorm is not found in the post-surgery model.'
+
+
+def test_low_precision_groupnorm_functional(device: str):
+    model = SimpleGroupNormConvModel()
```
Comment (dblalock) [inline tests/algorithms/test_low_precision_groupnorm.py:31]: if we're testing that it's "functional", should we do a forward + backward pass with it applied?

File: tests/algorithms/test_low_precision_groupnorm.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    assert model.modules is not None, 'model has .modules method'
+    for module_class in model.modules():
+        if isinstance(module_class, LayerNorm):
+            assert isinstance(
+                module_class,
+                LPGroupNorm), 'A standard torch.nn.LayerNorm should not be found in the model after surgery is applied.'
+
+    assert any(
+        isinstance(module_class, LPGroupNorm) for module_class in model.modules()
+    ), 'composer.algorithms.low_precision_groupnorm.low_precision_groupnorm.LPGroupNorm is not found in the post-surgery model.'
```
Comment (dblalock) [inline tests/algorithms/test_low_precision_groupnorm.py:27]:
```suggestion
    message = ('composer.algorithms.low_precision_groupnorm' + 
               '.low_precision_groupnorm.LPGroupNorm is not found ' +
               'in the post-surgery model.')
    assert any(isinstance(module_class, LPGroupNorm) for module_class in model.modules()), message
```

Or something like that. L27 is currently a 128-char line, which is will ugly line wrap in almost any editor (and technically violates our coding standard, but who knows what's up with yapf)

File: composer/algorithms/low_precision_layernorm/low_precision_layernorm.py [hunk 4]
```diff
@@ -99,6 +96,7 @@ def apply_low_precision_layernorm(model, optimizers: Union[torch.optim.Optimizer
         }
 
     replaced_instances = module_surgery.replace_module_classes(module=model, optimizers=optimizers, policies=policy)
+    print(f'Replaced {len(replaced_instances)} instances of LayerNorm with LowPrecisionLayerNorm')
```
Comment (dblalock) [inline composer/algorithms/low_precision_layernorm/low_precision_layernorm.py:99]: intentional print statement?

File: composer/algorithms/low_precision_groupnorm/metadata.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+    "low_precision_layernorm": {
+        "name": "Low Precision GroupNorm",
+        "class_name": "LowPrecisionGroupNorm",
+        "functional": "apply_low_precision_groupnorm",
+        "tldr": "Forces GroupNorm to fp16 or bf16",
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/metadata.json:6]:
```suggestion
        "tldr": "Forces GroupNorm to use fp16 or bf16",
```

File: composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    This algorithm is intended to be used instead of Fused GroupNorm. They have similar behavior and performance.
+
+    Args:
+        apply_at (Event, optional): Event where algorithm is applied.
+    """
+
+    def __init__(self, apply_at: Optional[Event] = None):
+        self.apply_at = Event.INIT if apply_at is None else apply_at
+        if self.apply_at not in {Event.INIT, Event.AFTER_LOAD}:
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py:90]: Is one of these two options a safe default? Would be nice to get have this algorithm work out of the box, but I know there are some thorny design issues here so no worries if that's infeasible.

File: composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        warnings.warn(NoEffectWarning('No instances of torch.nn.GroupNorm found.'))
+    log.info(f'Successfully replaced {len(replaced_instances)} instances of GroupNorm with LowPrecisionGroupNorm')
+
+
+class LowPrecisionGroupNorm(Algorithm):
+    """
+    Replaces all instances of `torch.nn.GroupNorm` with `composer.algorithms.low_precision_groupnorm.low_precision_groupnorm.LPGroupNorm`.
+
+    LPGroupNorm is a thin wrapper around `torch.nn.GroupNorm` which forces the layer to run in lower precision (torch.float16 or torch.bfloat16)
+    if autocast is enabled. This algorithm has no effect in FP32 or DeepSpeed FP16 mode, where autocast is disabled.
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py:80]:
```suggestion
    Replaces all instances of :class:`torch.nn.GroupNorm` with :class:`.LPGroupNorm`.

    `LPGroupNorm` is a thin wrapper around :class:`torch.nn.GroupNorm`_ which forces the layer to run in 
    lower precision (torch.float16 or torch.bfloat16) if autocast is enabled. This algorithm has no effect in 
    FP32 or DeepSpeed FP16 mode, where autocast is disabled.
```
(lines under 120 chars, internal links)

File: composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        warnings.warn(NoEffectWarning('Low Precision GroupNorm only applies to AMP_FP16 and AMP_BF16 precisions.'))
+        return model
+
+    policy: Dict[Type[torch.nn.Module], module_surgery.ReplacementFunction] = {torch.nn.GroupNorm: to_LPGroupNorm}
+
+    replaced_instances = module_surgery.replace_module_classes(module=model, optimizers=optimizers, policies=policy)
+    print(f'Replaced {len(replaced_instances)} instances of GroupNorm with LowPrecisionGroupNorm')
+    if len(replaced_instances) == 0:
+        warnings.warn(NoEffectWarning('No instances of torch.nn.GroupNorm found.'))
+    log.info(f'Successfully replaced {len(replaced_instances)} instances of GroupNorm with LowPrecisionGroupNorm')
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py:72]: unintentional print statement?

File: composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        downcast_x = _cast_if_autocast_enabled(x)
+        downcast_weight = _cast_if_autocast_enabled(self.weight)
+        downcast_bias = _cast_if_autocast_enabled(self.bias)
+        with torch.autocast(enabled=False, device_type=module_device.type):
+            return F.group_norm(downcast_x, self.num_groups, downcast_weight, downcast_bias, self.eps)
+
+
+def to_LPGroupNorm(layer: torch.nn.Module, module_index: int) -> LPGroupNorm:
+    assert isinstance(layer,
+                      torch.nn.GroupNorm), 'The replacement policy will look for all instances of torch.nn.GroupNorm'
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py:55]:
```suggestion
    message = 'The replacement policy will look for all instances of torch.nn.GroupNorm'
    assert isinstance(layer, torch.nn.GroupNorm), message
```
as long as it's two lines either way, might as well be prettier

File: composer/algorithms/low_precision_groupnorm/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+ `Natural Language Processing`, `Math Equivalent`
+
+Low Precision GroupNorm forces `torch.nn.GroupNorm` modules to run in float16 or bfloat16 precision, improving utilization. This should not affect final convergence values, but in rare cases may cause loss spikes.
+
+
+## How to Use
+Low Precision GroupNorm wraps `torch.nn.GroupNorm`, forcing the module to run in a lower precision if you have autocast enabled. If you are running in Automatic Mixed Precision (`amp`) mode, Low Precision GroupNorm will run in `torch.float16`. If you are running in `bf16` mode, Low Precision GroupNorm will run in `torch.bfloat16`.
+
+This algorithm will have no effect if you are running in `fp32` or `fp16` mode.
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/README.md:14]: how do I know what mode I'm running in? Is it just based on the precision I passed to the trainer? What if I'm not using the trainer? Am I in one of these modes by default?

(think 1-2 sentences could clarify all of this)
EDIT: or maybe just indirecting to the discussion of how to set the precision later in this page

File: composer/algorithms/low_precision_groupnorm/README.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+                  eval_dataloader=eval_dataloader,
+                  max_duration='1ep',
+                  algorithms=[LowPrecisionGroupNorm()])
+
+trainer.fit()
+```
+
+### Implementation Details
+
+Low Precision GroupNorm is implemented by performing model surgery, which looks for instances of `torch.nn.GroupNorm` and replaces them with `composer.algorithms.low_precision_groupnorm.low_precision_groupnorm.LPGroupNorm`, which is a thin wrapper around `torch.nn.GroupNorm` that manually turns autocast off and specifices the input dtype to lower precision.
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/README.md:69]:
```suggestion
Low Precision GroupNorm is implemented by performing model surgery, which looks for instances of `torch.nn.GroupNorm` and replaces them with `composer.algorithms.LPGroupNorm`, which is a thin wrapper around `torch.nn.GroupNorm` that manually turns autocast off and specifices the input dtype to lower precision.
```
should use the public import path we expect users to use

File: composer/algorithms/low_precision_groupnorm/README.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+## Suggested Hyperparameters
+
+Low Precision GroupNorm utilizes the existing parameters from the original model. The functional version of Low Precision GroupNorm allows users to specify the `precision` mode, which should be set to the Composer precision format of your model. When using the algorithm through the Composer trainer, Low Precision GroupNorm will use the trainer's `precision` mode.
+
+## Technical Details
+
+Low Precision GroupNorm is meant to replace our Fused GroupNorm algorithm. The two algorithms achieve very similar throughput. Fused GroupNorm also runs in low precision, but it is a more complex algorithm, since it uses a custom kernel. Since the custom kernel provides no additional speedup, we have replaced it with this simpler algorithm.
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/README.md:77]: We could add a "note" with the development history, but I probably don't care as a user. I'm expecting to see technical details about what it does in this section.

File: composer/algorithms/low_precision_groupnorm/__init__.py [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Replaces all instances of `torch.nn.GroupNorm` with a low precision `torch.nn.GroupNorm` (either float16 or bfloat16).
+By default, torch.autocast always runs torch.nn.GroupNorm in float32, so this surgery forces a lower precision.
```
Comment (dblalock) [inline composer/algorithms/low_precision_groupnorm/__init__.py:5]:
```suggestion
"""Replaces all instances of :class:`torch.nn.GroupNorm` with a low precision :class:`torch.nn.GroupNorm` (either float16 or bfloat16).
By default, `torch.autocast` always runs :class:`torch.nn.GroupNorm` in float32, so this surgery forces a lower precision.
```

**Review by dblalock — COMMENTED** [review]
I have no context on this PR, so I just went through and found ways to polish it.

File: tests/algorithms/test_low_precision_groupnorm.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+        run_name='run_name',
+        device=get_device(device),
+        dataloader=dataloader,
+        dataloader_label='train',
+        max_duration='1ep',
+    )
+    state._precision = Precision('amp_fp16')
+    low_precision_groupnorm = LowPrecisionGroupNorm()
+
+    low_precision_groupnorm.apply(Event.INIT, state, empty_logger)
```
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply tests/algorithms/test_low_precision_groupnorm.py:64]: Rejecting for readability, I think it's fine to add an extra line to be clear

File: tests/algorithms/test_low_precision_groupnorm.py (hunk 2 above)
Comment (dblalock) [reply tests/algorithms/test_low_precision_groupnorm.py]: I think there's enough value in it that it's worth doing at some point, but doesn't have to be in scope for this PR. Mostly I've just caught a lot of bugs in the old research repo + early composer by seeing if one forward + backward with the same data makes the loss go down.

**Comment by mvpatel2000** [comment]
Will switch to SimpleConvModel once https://github.com/mosaicml/composer/pull/1991 merges

**Review by dskhudia — APPROVED** [review]

File: composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+from composer.utils import module_surgery
+
+log = logging.getLogger(__name__)
+
+
+def _cast_if_autocast_enabled(hidden_states):
+    if not torch.is_autocast_enabled():
+        return hidden_states
+    else:
+        return torch.cuda.amp.autocast_mode._cast(hidden_states, torch.get_autocast_gpu_dtype())
```
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply composer/algorithms/low_precision_groupnorm/low_precision_groupnorm.py:28]: I have manually verified it. I am having trouble w export tests so I am skipping adding a unit test as we need this in soon

**Comment by mvpatel2000** [comment]
I need a different logo... any suggestions for icon

File: composer/algorithms/low_precision_layernorm/low_precision_layernorm.py [hunk 16]
```diff
@@ -72,18 +70,17 @@ def to_LPLayerNorm(layer: torch.nn.Module, module_index: int) -> LPLayerNorm:
 
 def to_FusedLayerNorm(layer: torch.nn.Module, module_index: int) -> APEXFusedLayerNorm:
     """Defines a replacement policy from a `torch.nn.LayerNorm` to a `apex.normalization.fused_layer_norm`"""
-    assert isinstance(layer,
-                      torch.nn.LayerNorm), 'The replacement policy will look for all instances of torch.nn.LayerNorm'
+    assert isinstance(layer, torch.nn.LayerNorm), 'Replacement policy only replaces torch.nn.LayerNorm'
```
Comment (dblalock) [inline composer/algorithms/low_precision_layernorm/low_precision_layernorm.py:73]: technically this should be a TypeError since [asserts are more for detecting bugs / documenting program logic, not raising user-visible exceptions](https://stefan.sofa-rockers.org/2018/04/16/assertions-and-exceptions/)

File: composer/algorithms/low_precision_layernorm/low_precision_layernorm.py [hunk 17]
```diff
@@ -106,19 +103,20 @@ def apply_low_precision_layernorm(model, optimizers: Union[torch.optim.Optimizer
...
 
     This algorithm is intended to be used instead of Fused LayerNorm. They have similar behavior and performance.
 
     Args:
         apply_at (Event, optional): Event where algorithm is applied.
     """
 
-    def __init__(self, apply_at: Optional[Event] = None):
-        self.apply_at = Event.INIT if apply_at is None else apply_at
+    def __init__(self, apply_at: Optional[Event] = Event.INIT):
```
Comment (dblalock) [inline composer/algorithms/low_precision_layernorm/low_precision_layernorm.py:118]: Do we actually want this to be Optional? Is there a use case for passing `None`?

File: tests/algorithms/test_low_precision_groupnorm.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+from tests.common import RandomImageDataset
+from tests.common.models import SimpleConvModel
+
+
+def assert_is_lpgn_instance(model):
+    # ensure that within the entire model, no PyTorch GroupNorm exists, and at least one LPGN does.
+    assert model.modules is not None, 'model has .modules method'
+    for module_class in model.modules():
+        if isinstance(module_class, LayerNorm):
+            assert isinstance(module_class, LPGroupNorm)
```
Comment (dblalock) [inline tests/algorithms/test_low_precision_groupnorm.py:21]: typo?
```suggestion
        if isinstance(module_class, GroupNorm):
            assert isinstance(module_class, LPGroupNorm)
```

**Review by dblalock — APPROVED** [review]
basically LGTM, except maybe one test needs to say "GroupNorm" instead of "LayerNorm"
